icu4c/source/common/umutex.h

```diff
@@ -186,30 +187,45 @@ template<class T> void umtx_initOnce(UInitOnce &uio, void (U_CALLCONV *fp)(T, UE
  * platform independent set of mutex operations.  Now vestigial, wrapping std::mutex only.
  * For internal ICU use only.
  *
- * Caution: do not directly declare static or global instances of UMutex. Doing so can introduce
- * static initializers, which are disallowed in ICU library code. Instead, use the following
- * idiom, which avoids static init and also avoids ordering issues on destruction
- * (use after delete) by avoiding destruction altogether.
- *
- *  UMutex *myMutex() {
- *      static UMutex *m = STATIC_NEW(UMutex);
- *      return m;
- *  }
- *  ...
- *
- *  Mutex lock(myMutex());   // hold myMutex until the variable "lock" goes out of scope.
+ * TODO: redo this doc section.
  */
 
-struct UMutex : public icu::UMemory {
+class UMutex {
+public:
     UMutex() = default;
     ~UMutex() = default;
+
     UMutex(const UMutex &other) = delete;
     UMutex &operator =(const UMutex &other) = delete;
+    void *operator new(size_t) = delete;
+
+    // requirements for C++ BasicLockable, allows UMutex to work with std::lock_guard
+    void lock() {
+        std::mutex *m = fMutex.load(std::memory_order_acquire);
+        if (m == nullptr) { m = getMutex(); }
+        m->lock();
+    }
+    void unlock() { fMutex.load(std::memory_order_relaxed)->unlock(); };
 
-    std::mutex   fMutex = {};    // Note: struct - pubic members - because most access is from
-    //                           //       plain C style functions (umtx_lock(), etc.)
+    static void cleanup();
+
+private:
+    alignas(std::mutex) char fStorage[sizeof(std::mutex)];
+    std::atomic<std::mutex *> fMutex;
+
+    /** All initialized UMutexes are kept in a linked list, so that they can be found,
```

Comment:
Thank you Andy! This is a neat solution, so that they will also be destroyed in the order they were created too.